### ISSUE TITLE
Returning optional union from method complains about `got "object"`

### ISSUE THREAD

**Issue by Garrett-R** [python/mypy#12823]
**Bug Report**

Mypy thinks my method returns "object" which is not right.

**To Reproduce**

Run Mypy on this file:

```python
from typing import Optional


class Cat:
    pass


class Dog:
    pass


class MyClass:
    def __init__(self) -> None:
        self.pet: Optional[Dog | Cat] = None

    def get_pet(self) -> Optional[Dog | Cat]:
        if input() == 'yes':
            self.pet = Cat()
        else:
            self.pet = Dog()
        return self.pet  # (incorrect Mypy error on this line!)
```

**Expected Behavior**

This should have no errors.

**Actual Behavior**

It yields:

```
error: Incompatible return value type (got "object", expected "Union[Dog, Cat, None]")
```

**Your Environment**

- Mypy version used: v0.950
- Mypy command-line flags:  _none_
- Mypy configuration options: _none_
- Python version used: 3.10.4
- Operating system and version: Ubuntu 20.04

**Comment by erictraut** [comment]
Looks like another case where a false positive results from using a join operation rather than a union.

**Comment by hauntsaninja** [comment]
Note the following works:
```
    def get_pet(self) -> Dog | Cat | None:
        pet: Dog | Cat | None = None
        if input() == 'yes':
            pet = Cat()
        else:
            pet = Dog()
        return pet
```
Looks like we're losing the type context of the instance variable.

**Comment by sterliakov** [comment]
Fixed in #18538 and passes on current master.
